internal/rotate_est.go

```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return fmt.Errorf("Unable to serialize new private key: %w", err)
+		}
+		keyBytes = pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: keyBytes})
+		newKey = string(keyBytes)
+		signer = key
+	}
+
+	// Ask EST server for new cert
+	csrBytes, err := createB64CsrDer(signer, cert)
+	if err != nil {
+		return err
+	}
+
+	url := handler.State.EstServer + "/simplereenroll"
+	res, err := handler.client.Post(url, "application/pkcs10", bytes.NewBuffer(csrBytes))
+	if err != nil {
+		return fmt.Errorf("Unable to submit certificate signing request: %w", err)
+	}
+	buf, err := io.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Errorf("Unable to read certificate response body: HTTP_%d - %w", res.StatusCode, err)
+	}
+	if res.StatusCode != 201 {
+		return fmt.Errorf("Unable to obtain new certificate: HTTP_%d - %s", res.StatusCode, string(buf))
+	}
+	ct := res.Header.Get("content-type")
+	if ct != "application/pkcs7-mime" {
+		return fmt.Errorf("Unexpected content-type return in certificate response: %s", ct)
+	}
+	estCert, err := decodeEstResponse(string(buf))
+	if err != nil {
+		return err
+	}
+
+	// Do minimal sanity checking on the new cert
+	if err = verifyNewCert(cert, estCert); err != nil {
+		return err
+	}
+
+	// Update our state
+	if handler.usePkcs11() {
+		newCert := "09"
+		if err = handler.crypto.ctx.ImportCertificate(idToBytes(newCert), estCert); err != nil {
+			return fmt.Errorf("Unable to import new cert into HSM: %w", err)
+		}
+		handler.State.NewCert = newCert
+	} else {
+		handler.State.NewCert = string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: estCert.Raw}))
+	}
+
+	handler.State.NewKey = newKey
+	return nil
+}
+
+// createB64CsrDer creates the payload for an EST simplereenroll payload. The
+// main thing the EST server will want is for the x509 subject to be the same.
+// Then we also need to ask for the proper x509 extensions.
+func createB64CsrDer(key crypto.Signer, cert *x509.Certificate) ([]byte, error) {
+	template := x509.CertificateRequest{
+		PublicKeyAlgorithm: 0,
+		PublicKey:          key.Public(),
+		RawSubject:         cert.RawSubject,
+		ExtraExtensions: []pkix.Extension{
+			{
+				Id:       oidKeyUsage,
+				Critical: true,
+				Value:    asn1DigitalSignature,
+			},
+			{
+				Id:       oidExtendedKeyUsage,
+				Critical: true,
+				Value:    asn1TlsWebClientAuth,
+			},
+		},
+	}
+
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, &template, key)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(base64.StdEncoding.EncodeToString(csrBytes)), nil
+}
+
+func decodeEstResponse(estResponse string) (*x509.Certificate, error) {
+	bytes, err := base64.StdEncoding.DecodeString(estResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to base64 decode EST response: %w", err)
+	}
+	p7, err := pkcs7.Parse(bytes)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid pkcs7 data in EST response: %w", err)
+	}
+	return p7.Certificates[0], nil
```

Comment:
I suppose the previous check guarantee that `len(p7.Certificates) > 0`?

Reply:
Correct. If there's no err, there must have been a response.